### PR TITLE
Create Room Manager

### PR DESCRIPTION
**Pull request by Alton09**
## Description

First iteration of decoupling Video SDK logic from the RoomActivity. The RoomManager is currently housing all of the video SDK view state management. Ideally the view state should be managed in a separate [ViewModel](https://developer.android.com/reference/androidx/lifecycle/ViewModel.html) class, but that can be done in a future story to avoid scope creep in this story.

Note that the RoomViewEffect, RoomViewState, and RoomViewEvent classes along with the updateState, viewEffect, and withState functions are all part of the MVVM pattern. Once we pull view state management into the view model we can abstract these functions and make them reusable for future screens.

## Breakdown

- Create RoomManager to encapsulate Room management.
- Temporarily removed lint check since much of the existing RoomActivity code has been temporarily commented out.
- Add static factory function to GoogleAuthProvider.
- Remove SharedPreferences dependency from AuthProviders.
- Remove redundant audi/video toggle menu items.

## Validation

- Manually tested local video track rendering, connecting to a room with other participants, and toggling audio.

## Additional Notes

N/A

## Submission Checklist

 - [x] The `CHANGELOG.md` reflects any **feature**, **bug fixes**, or **known issues** made in the source code

### CODE REVIEW

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java [hunk 1]
```diff
@@ -281,32 +180,48 @@ protected void onCreate(Bundle savedInstanceState) {
...
         requestPermissions();
     }
 
     @Override
     protected void onResume() {
         super.onResume();
-        boolean isAppLinkProvided = checkIntentURI();
-        displayName = sharedPreferences.getString(Preferences.DISPLAY_NAME, null);
-        updateUi(room, isAppLinkProvided);
+        //        boolean isAppLinkProvided = checkIntentURI();
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Let's add a checklist to the PR to the items that need to be re-enabled.

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java [hunk 2]
```diff
@@ -281,32 +180,48 @@ protected void onCreate(Bundle savedInstanceState) {
...
+    }
+
+    private void bindViewEffects(@Nullable RoomViewEffect viewEffect) {
+        if (viewEffect != null) {
+            if (viewEffect instanceof ScreenShareError) {
+                Snackbar.make(primaryVideoView, R.string.screen_capture_error, Snackbar.LENGTH_LONG)
+                        .show();
+            }
+            if (viewEffect instanceof RequestScreenSharePermission) {
+                //                requestScreenCapturePermission();
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Yeah same thing here. We should either cut tickets or track these in a PR description checklist so what needs to be completed for this work to be merged.

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java [hunk 3]
```diff
@@ -384,50 +277,43 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
         switchCameraMenuItem = menu.findItem(R.id.switch_camera_menu_item);
+        screenCaptureMenuItem = menu.findItem(R.id.share_screen_menu_item);
         pauseVideoMenuItem = menu.findItem(R.id.pause_video_menu_item);
         pauseAudioMenuItem = menu.findItem(R.id.pause_audio_menu_item);
-        screenCaptureMenuItem = menu.findItem(R.id.share_screen_menu_item);
 
         // Screen sharing only available on lollipop and up
         screenCaptureMenuItem.setVisible(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP);
 
+        roomManager.getViewState().observe(this, this::bindViewState);
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: This is an odd place to observe the LiveData. Why was this moved from onCreate?

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java [hunk 4]
```diff
@@ -770,68 +484,15 @@ private boolean permissionsGranted() {
...
-
-        primaryVideoView.showIdentityBadge(false);
+    private void updateUi(RoomViewState viewState) {
+        updateUi(viewState, false);
     }
 
-    private void updateUi(Room room) {
-        updateUi(room, false);
+    private void updateUi(boolean isAppLinkProvided) {
+        updateUi(null, isAppLinkProvided);
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: I don't think we should allow a null RoomViewState to be passed to updateUi. We should be able to get an initial RoomViewState from the RoomManager even when the activity is launched as a result of a deep link click

File: app/src/main/java/com/twilio/video/app/videosdk/RoomManager.kt [hunk 5]
```diff
@@ -0,0 +1,899 @@
+package com.twilio.video.app.videosdk
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/videosdk/RoomManager.kt]: We should just keep these new classes under the `com.twilio.video.app.ui.room`

This is a similar approach to the io app

https://github.com/google/iosched/blob/master/mobile/src/main/java/com/google/samples/apps/iosched/ui/signin/SignInViewModelDelegate.kt

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java (hunk 1 above)
Comment (Alton09) [reply app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Will create follow up tickets that will encompass what is left to be re-enabled.

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java (hunk 3 above)
Comment (Alton09) [reply app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Yeah I don't like it here either, but when binding the menu items for audio & video toggling, a NPE will get thrown since onCreateOptionsMenu gets invoked after onCreate.

Comment (Alton09) [reply app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Another solution to this is to add another livedata observer function just for menu items. Might be a bit better, but downside is the view state will be bound in two functions which reduces readability of the class.

File: app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java (hunk 4 above)
Comment (Alton09) [reply app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java]: Good catch! I went ahead and removed this overloaded method. The RoomManager updateState function guards against null RoomViewState values and also the MutableLiveData viewState property initializes with a default RoomViewState instance.

File: app/src/main/java/com/twilio/video/app/videosdk/RoomManager.kt (hunk 5 above)
Comment (Alton09) [reply app/src/main/java/com/twilio/video/app/videosdk/RoomManager.kt]: Yep that's fine for now, but I'd argue package by feature is the way to go opposed to packaging by layer. [This quick article](https://hackernoon.com/package-by-features-not-layers-2d076df1964d) explains the benefits. Also, later down the road if we want to modularize the app, each feature package could become its own module and will reduce the dependencies from other modules.
